File: packages/sdk/src/resources/marketplace.ts
/**
 * Marketplace resource client — public product listings + discovery.
 *
 * Maps to `/v1/marketplace`.
 */
import type { MarketplaceListing } from "@settlekit/common";
import type { HttpClient, RequestOptions } from "../http-client.js";

/** Input for {@link MarketplaceResource.createListing}. */
export interface CreateListingInput {
  organizationId: string;
  merchantId: string;
  productId: string;
  title: string;
  summary: string;
  tags?: string[];
}

/** Discovery query for {@link MarketplaceResource.search}. */
export interface ListingSearch {
  query?: string;
  tag?: string;
  sort?: "top" | "new" | "price";
}

/** Aggregate reputation/inventory view of a seller. */
export interface SellerProfile {
  merchantId: string;
  totalListings: number;
  publishedListings: number;
  totalRatings: number;
  ratingAverage: number;
  listingIds: string[];
}

/** Client for marketplace endpoints. */
export class MarketplaceResource {
  constructor(private readonly http: HttpClient) {}

  /** Create an (unpublished) listing for a product. */
  createListing(input: CreateListingInput, options?: RequestOptions): Promise<MarketplaceListing> {
    return this.http.post<MarketplaceListing>("/v1/marketplace/listings", input, options);
  }

  /** Search published listings. */
  search(query: ListingSearch = {}, options?: RequestOptions): Promise<MarketplaceListing[]> {
    const params = new URLSearchParams();
    if (query.query) params.set("q", query.query);
    if (query.tag) params.set("tag", query.tag);
    if (query.sort) params.set("sort", query.sort);
    const qs = params.toString();
    return this.http.get<MarketplaceListing[]>(`/v1/marketplace/listings${qs ? `?${qs}` : ""}`, options);
  }

  /** Retrieve a listing by id. */
  retrieve(id: string, options?: RequestOptions): Promise<MarketplaceListing> {
    return this.http.get<MarketplaceListing>(`/v1/marketplace/listings/${encodeURIComponent(id)}`, options);
  }

  /** Publish a listing (make it discoverable). */
  publish(id: string, options?: RequestOptions): Promise<MarketplaceListing> {
    return this.http.post<MarketplaceListing>(`/v1/marketplace/listings/${encodeURIComponent(id)}/publish`, undefined, options);
  }

  /** Remove a listing from discovery. */
  unpublish(id: string, options?: RequestOptions): Promise<MarketplaceListing> {
    return this.http.post<MarketplaceListing>(`/v1/marketplace/listings/${encodeURIComponent(id)}/unpublish`, undefined, options);
  }

  /** Add a 1–5 star rating to a listing. */
  rate(id: string, stars: number, options?: RequestOptions): Promise<MarketplaceListing> {
    return this.http.post<MarketplaceListing>(`/v1/marketplace/listings/${encodeURIComponent(id)}/rate`, { stars }, options);
  }

  /** Fetch a seller's aggregate profile. */
  seller(merchantId: string, options?: RequestOptions): Promise<SellerProfile> {
    return this.http.get<SellerProfile>(`/v1/marketplace/sellers/${encodeURIComponent(merchantId)}`, options);
  }
}

File: sdks/rust/src/resources/marketplace.rs
//! Marketplace resource (`/v1/marketplace`) — public listings + discovery.

use reqwest::Method;
use serde::Serialize;

use crate::{Client, Result};

/// Accessor for marketplace operations.
#[derive(Debug)]
pub struct Marketplace<'a> {
    client: &'a Client,
}

/// Body for creating a listing (`POST /v1/marketplace/listings`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateListing {
    pub organization_id: String,
    pub merchant_id: String,
    pub product_id: String,
    pub title: String,
    pub summary: String,
    #[serde(default)]
    pub tags: Vec<String>,
}

impl<'a> Marketplace<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        Marketplace { client }
    }

    /// Create an (unpublished) listing for a product.
    pub async fn create_listing(&self, body: &CreateListing) -> Result<serde_json::Value> {
        self.client
            .request(Method::POST, "/v1/marketplace/listings", Some(body))
            .await
    }

    /// Search published listings. `sort` is one of `top` | `new` | `price`.
    pub async fn search(
        &self,
        query: Option<&str>,
        tag: Option<&str>,
        sort: Option<&str>,
    ) -> Result<serde_json::Value> {
        let mut params: Vec<String> = Vec::new();
        if let Some(q) = query {
            params.push(format!("q={q}"));
        }
        if let Some(t) = tag {
            params.push(format!("tag={t}"));
        }
        if let Some(s) = sort {
            params.push(format!("sort={s}"));
        }
        let path = if params.is_empty() {
            "/v1/marketplace/listings".to_string()
        } else {
            format!("/v1/marketplace/listings?{}", params.join("&"))
        };
        self.client.request_no_body(Method::GET, &path).await
    }

    /// Retrieve a listing by id.
    pub async fn get(&self, id: &str) -> Result<serde_json::Value> {
        self.client
            .request_no_body(Method::GET, &format!("/v1/marketplace/listings/{id}"))
            .await
    }

    /// Publish a listing (make it discoverable).
    pub async fn publish(&self, id: &str) -> Result<serde_json::Value> {
        self.client
            .request_no_body(
                Method::POST,
                &format!("/v1/marketplace/listings/{id}/publish"),
            )
            .await
    }

    /// Add a 1–5 star rating to a listing.
    pub async fn rate(&self, id: &str, stars: u8) -> Result<serde_json::Value> {
        #[derive(Serialize)]
        struct Rate {
            stars: u8,
        }
        self.client
            .request(
                Method::POST,
                &format!("/v1/marketplace/listings/{id}/rate"),
                Some(&Rate { stars }),
            )
            .await
    }

    /// Fetch a seller's aggregate profile.
    pub async fn seller(&self, merchant_id: &str) -> Result<serde_json::Value> {
        self.client
            .request_no_body(
                Method::GET,
                &format!("/v1/marketplace/sellers/{merchant_id}"),
            )
            .await
    }
}
